### PR TITLE
TIP-668: introduce a Product Value Collection

### PR DESCRIPTION
**Pull request by jjanvier**
This PR introduces a "real business" collection for product values.

This improves the following:
- we are now sure we manipulate `Pim\Component\Catalog\Model\ProductValueInterface` objects in the collection
- we now ensure the values normalizer only supports this collection (instead of all Doctrine collections)
- the indexation of the value is now internal to the collection, and not exposed in the product itself. (btw, the indexation is here to be able to efficiently retrieve a product value, without iterating on all existing values)
- remove the indexation that was made when Doctrine loaded the collection (as the values are not handled by Doctrine anymore now)

At first I wanted to use simply a Doctrine `ArrayCollection`. But it's not doable as we need our own indexation and the `$lements` property of the `ArrayCollection` is private.
Then I implemented the Doctrine `Collection` interface instead of doing our own interface. Two drawbacks:
- it implements `ArrayAccess` which means we can choose the index of the value, which would break the internal indexation
- it has many many methods we don't use often, and I didn't want to so much code for nothing. If we need other methods later (like filtering for instance), we'll do that later.

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | -
| Added integration tests           | -
| Changelog updated                 | Y
| Review and 2 GTM                  | :clock1:
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | not yet

### CODE REVIEW

File: src/Pim/Component/Catalog/Model/AbstractProduct.php [hunk 1]
```diff
@@ -340,35 +243,13 @@ public function setRawValues(array $rawValues)
...
-     */
-    public function __isset($attributeCode)
-    {
-        $indexedValues = $this->getIndexedValues();
+        $attributes = [];
 
-        return isset($indexedValues[$attributeCode]);
-    }
+        foreach ($this->values as $value) {
+            $attributes[] = $value->getAttribute();
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Model/AbstractProduct.php]: dupplicates here

File: src/Pim/Component/Catalog/Model/ProductValueCollection.php [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+ * to ensure the inner indexation by attribute-channel-locale.
+ *
+ * Ideally, we should not rely on Doctrine collections. But hey, ¯\_(ツ)_/¯, nothing's perfect,
+ * and it's already much better that what we had :)
+ *
+ * @author    Julien Janvier <j.janvier@gmail.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ProductValueCollection implements Collection, Selectable
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Model/ProductValueCollection.php]: implements Countable + IteratorAggregate + only what's useful maybe

File: src/Pim/Component/Catalog/Model/AbstractProduct.php (hunk 1 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/AbstractProduct.php]: done

**Comment by jjanvier** [comment]
I will squash commits during merge

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php [hunk 3]
```diff
@@ -0,0 +1,168 @@
+<?php
+
+namespace Pim\Component\Catalog\Model;
+
+/**
+ * Business collection interface to handle product values.
+ * It does not extends \IteratorAggregate as indexation could be performed internally
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: So does it extends or not? Here you say "does not", but below you have `extends \Countable, \IteratorAggregate`

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+ * Highly inspired by \Doctrine\Common\Collection.
+ *
+ * @author    Julien Janvier <j.janvier@gmail.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface ProductValueCollectionInterface extends \Countable, \IteratorAggregate
+{
+    /**
+     * Adds an value at the end of the collection.
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: a value

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+     */
+    public function add(ProductValueInterface $value);
+
+    /**
+     * Clears the collection, removing all values.
+     */
+    public function clear();
+
+    /**
+     * Checks whether an value is contained in the collection.
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: a value

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+     * Removes the specified value from the collection, if it is found.
+     *
+     * @param ProductValueInterface $value The value to remove.
+     *
+     * @return bool TRUE if this collection contained the specified value, FALSE otherwise.
+     */
+    public function remove(ProductValueInterface $value);
+
+    /**
+     * Checks whether the collection contains an value with the specified key/index.
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: a value

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+     *
+     * @param string|integer $key The key/index to check for.
+     *
+     * @return bool TRUE if the collection contains an value with the specified key/index,
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: a value

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+     *                 FALSE otherwise.
+     */
+    public function containsKey($key);
+
+    /**
+     * Gets the value at the specified key/index.
+     *
+     * @param string $key The key/index of the value to retrieve.
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php:80]: "string|int" as above? And "int" or "integer"

File: src/Pim/Component/Catalog/Model/ProductValueCollection.php [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    {
+        return current($this->values);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function removeKey($key)
+    {
+        if ( ! isset($this->values[$key]) && ! array_key_exists($key, $this->values)) {
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/ProductValueCollection.php]: no spaces → `!isset()` (php-cs-fixer is unhappy).

File: src/Pim/Component/Catalog/Model/ProductValueCollection.php [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function containsKey($key)
+    {
+        return isset($this->values[$key]) || array_key_exists($key, $this->values);
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/ProductValueCollection.php]: Why do you need both tests? Why not just array_key_exists, as we only look for the key and don't care about the value?

**Review by damien-carcel — APPROVED** [review]
Minor comments on PHPDoc, otherwise GTM.

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php (hunk 3 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: I meant ArrayAccess
I fix

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php (hunk 4 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: done

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php (hunk 5 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: done

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php (hunk 6 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: done

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php (hunk 8 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php:80]: if fixed `containsKey` above
the indexation is made internally and the indexes can only be strings

File: src/Pim/Component/Catalog/Model/ProductValueCollection.php (hunk 10 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductValueCollection.php]: done

File: src/Pim/Component/Catalog/Model/ProductValueCollection.php (hunk 9 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductValueCollection.php]: done

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+     * @return bool TRUE if the collection is empty, FALSE otherwise.
+     */
+    public function isEmpty();
+
+    /**
+     * Removes the value at the specified index from the collection.
+     *
+     * @param string $key The kex/index of the value to remove.
+     *
+     * @return ProductValueInterface The removed value or NULL, if the collection did not contain the value.
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: ProductValueInterface|null

File: src/Pim/Component/Catalog/Model/ProductValueCollection.php [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+    {
+        return current($this->values);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function removeKey($key)
+    {
+        if (!isset($this->values[$key]) && !array_key_exists($key, $this->values)) {
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Model/ProductValueCollection.php]: array_key_exists should be enough, no ?

File: src/Pim/Component/Catalog/Model/ProductValueCollection.php [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function remove(ProductValueInterface $value)
+    {
+        $key = array_search($value, $this->values, true);
+
+        if ($key === false) {
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Model/ProductValueCollection.php]: yoda :)

**Review by skeleton — APPROVED** [review]
baby comments, nice work ;)
GTM

File: src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php (hunk 11 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductValueCollectionInterface.php]: done

File: src/Pim/Component/Catalog/Model/ProductValueCollection.php (hunk 12 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductValueCollection.php]: done

File: src/Pim/Component/Catalog/Model/ProductValueCollection.php (hunk 13 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductValueCollection.php]: done

**Pull request merged**